implementations/rust/ockam/ockam_api/src/kafka/kafka_portal_listener.rs

```diff
@@ -0,0 +1,85 @@
+use crate::kafka::kafka_inlet_map::KafkaInletMap;
+use crate::kafka::kafka_portal_worker::KafkaPortalWorker;
+use ockam::Context;
+use ockam_core::{Address, AllowAll, Any, Routed, Worker};
+
+
+
+use std::sync::Arc;
+
+use tracing::trace;
+use tracing::warn;
+
+///Acts like an outlet listener of tcp transport
+/// only receives first ping and spawn a dedicated worker to handle the rest
+pub(crate) struct KafkaPortalListener {
+    inlet_map: Arc<KafkaInletMap>,
+}
+
+#[ockam::worker]
+impl Worker for KafkaPortalListener {
+    type Message = Any;
+    type Context = Context;
+
+    async fn handle_message(
+        &mut self,
+        context: &mut Self::Context,
+        message: Routed<Self::Message>,
+    ) -> ockam::Result<()> {
+        trace!("received message: {:?}", &message);
+
+        let result = KafkaPortalWorker::start(context, self.inlet_map.clone()).await;
+        if let Err(err) = result {
+            warn!("cannot spawn worker; err = {:?}", err);
+            return Err(err);
+        }
+        let worker_address = result.unwrap();
+
+        //forward to the worker and place its address in the return route
+        let mut message = message.into_local_message();
+
+        message
+            .transport_mut()
+            .onward_route
+            .modify()
+            .replace(worker_address.clone());
+
+        trace!(
+            "forwarding message: onward={:?}; return={:?}; worker={:?}",
+            &message.transport().onward_route,
+            &message.transport().return_route,
+            worker_address
+        );
+
+        context.forward(message).await?;
+
+        Ok(())
+    }
+}
+
+impl KafkaPortalListener {
+    pub(crate) async fn bootstrap(
+        context: &Context,
+        portal_address: Address,
+        bind_address: String,
+        port_range: (u16, u16),
+    ) -> ockam_core::Result<()> {
+        Self::start(
+            context,
+            portal_address,
+            Arc::new(KafkaInletMap::new(bind_address, port_range)),
+        )
+        .await
+    }
+
+    pub(crate) async fn start(
+        context: &Context,
+        portal_address: Address,
+        inlet_map: Arc<KafkaInletMap>,
+    ) -> ockam_core::Result<()> {
+        //TODO: only allow the inlet
+        context
+            .start_worker(portal_address, Self { inlet_map }, AllowAll, AllowAll)
```

Comment:
Ping me if you need some help with figuring out AccessControl here

Reply:
Yep, I need help, I tried to naively allow only inlets listeners, but what I actually need are the TCP inlet workers instead. 

https://github.com/build-trust/ockam/blob/davide-baldo/kafka-sidecar/implementations/rust/ockam/ockam_api/src/kafka/kafka_portal_listener.rs#L64

Reply:
removed the code, I will fix permissions on the next iteration